packages/unmock-core/src/service/util.ts

```diff
@@ -1,6 +1,51 @@
+import fs from "fs";
+import pointer from "json-pointer";
+import fspath from "path";
 import XRegExp from "xregexp";
 import { OAS_PATH_PARAM_REGEXP, OAS_PATH_PARAMS_KW } from "./constants";
-import { Parameter, Paths } from "./interfaces";
+import { OpenAPIObject, Parameter, Paths } from "./interfaces";
+
+/**
+ * Simple dereferencing JSON schema $ref with JSON pointers or URIs.
+ * See https://json-schema.org/understanding-json-schema/structuring.html for more.
+ */
+export function derefIfNeeded(
+  mightHaveReference: any,
+  derefSchema: { schema: OpenAPIObject; absPath: string },
+): any {
+  if (
+    typeof mightHaveReference !== "object" ||
+    Object.keys(mightHaveReference).length === 0
+  ) {
+    return mightHaveReference;
+  }
+  // DFS deref all items as needed
+  for (const childKey of Object.keys(mightHaveReference)) {
+    mightHaveReference[childKey] = derefIfNeeded(
+      mightHaveReference[childKey],
+      derefSchema,
+    );
+  }
+  const refValue = mightHaveReference.$ref;
+  if (refValue === undefined) {
+    return mightHaveReference;
+  }
+  // decide between local and path URI
+  const isLocal = refValue.startsWith("#");
+  const afterPound = refValue
+    .split("#")
+    .slice(1)
+    .join("#");
+
+  const schema = isLocal
+    ? derefSchema.schema
+    : fs.readFileSync(
+        fspath.join(derefSchema.absPath, refValue.split("#")[0]),
```

Comment:
Should one perhaps check if the file referenced by the reference even exists? Could it be something else than a file as well?

Reply:
It can most definitely be something other than a file. In theory, it could also be e.g. a web URL, a link to `$id`, etc. There are many options and I'm not sure if we want to implement all of them for now - or at least in this PR. They could also be a subject of different "dereferencers". I can see that being set in e.g. `UnmockOptions`.

If the file does not exist - it will throw and I think it's a quite verbose error when a file is not found.

Reply:
Ok. I don't really understand how different dereferencers would work, afaics there's only one way to resolve a reference and that's by following the reference. Also don't see why it needs to be mocked 🤔 I can see issues where the dereferencing follows external URLs but not sure if that would be easier to mock by patching instead of passing the dereferencer around...

Reply:
In theory, one could pass e.g. `offlineMode` as parameters and then fail when trying to resolve URLs fetches. But like I said - I don't _really_ expect different kinds of dereferencers.

It only needs to be mocked when e.g. testing a `State` that was not initialized by a `Service` (then the `dereferencer` is not initialized, but `state.update` still expects it). Same when checking other methods that use a `dereferencer`.

Reply:
Cool. I kinda wouldn't have expected it to be a service's job to keep track of how references should be resolved for other users of the schema but maybe exposing a decorated schema that is aware of how to resolve those. But I don't understand the full architecture well enough to know which one is best, does not really matter.